danceschool/core/handlers.py

```diff
@@ -1,89 +1,101 @@
+        return
+
+    user = email_address.user
+
+    if not hasattr(user, 'customer'):
+        last_reg = Registration.objects.filter(
+            customer__email=email_address.email,
+            customer__user__isnull=True,
+            dateTime__isnull=False
+        ).order_by('-dateTime').first()
+
+        if last_reg:
+            customer = last_reg.customer
+            customer.user = user
+            customer.save()
+
+            if not user.first_name and not user.last_name:
+                user.first_name = customer.first_name
+                user.last_name = customer.last_name
+                user.save()
+
+
+@receiver(post_registration)
+def linkCustomerToVerifiedUser(sender, **kwargs):
+    """
+    If a Registration is processed in which the associated Customer does not yet
+    have a User, then check to see if the Customer's email address has been
+    verified as belonging to a specific User, and if that User has an associated
+    Customer.  If such a User is found, then associated this Customer with that
+    User.  This way, if a new User verifies their email account before they have
+    submitted any Registrations, their Customer account is seamlessly linked when
+    they do complete their first Registration.
+    """
+    registration = kwargs.get('registration', None)
+
+    if not registration or (hasattr(registration.customer,'user') and registration.customer.user):
+        return
+
+    logger.debug('Checking for User for Customer with no associated registration.')
+
+    customer = registration.customer
+
+    try:
+        verified_email = EmailAddress.objects.get(
+            email=customer.email,
+            verified=True,
+            primary=True,
+            user__customer__isnull=True
+        )
+
+        logger.info("Found user {} to associate with customer {}.".format(verified_email.user.id, customer.id))
+
+        customer.user = verified_email.user
+        customer.save()
+
+        if not customer.user.first_name and not customer.user.last_name:
+            customer.user.first_name = customer.first_name
+            customer.user.last_name = customer.last_name
+            customer.user.save()
+    except ObjectDoesNotExist:
+        logger.info("No user found to associate with customer {}.".format(customer.id))
+    except MultipleObjectsReturned as e:
+        # This should never happen, as email should be unique in the db table account_emailaddress.
+        # If it does, something's broken in the database or Django.
+        logger.error(
+            "Something's not right with the database: more than one entry found on the database for the email {}. "
+            "This duplicate key value violates unique constraint \"account_emailaddress_email_key\". "
+            "The email field should be unique for each account.".format(customer.email))
+        logger.error("Exception:")
+        logger.error(e)
```

Comment:
IMO each logging call should be concerned with one message.

You can embed `\n` in the text to have `Exception:` on its own line, and use `logger.error(msg, exc_info=True)` or its equivalent `logger.exception(msg)` to have the traceback included (which is longer but much more useful than only `str(exc)`)

Reply:
Implemented.